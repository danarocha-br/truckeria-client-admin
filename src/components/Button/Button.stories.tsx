import { Story } from '@storybook/react/types-6-0';
import { FiPlus } from 'react-icons/fi';

import Button from '.';

export default {
  title: 'Button',
  component: Button,
  args: {
    label: 'Sign in',
    ariaLabel: 'View more',
    kind: 'primary',
    size: 'md',
    fullWidth: false,
    disabled: false,
    loading: false,
    onClick: () => 'clicked',
  },
  argTypes: {
    label: {
      table: {
        category: 'Text',
      },
    },
    ariaLabel: {
      table: {
        category: 'Text',
      },
    },
    onClick: {
      table: {
        category: 'Events',
      },
    },
    variant: {
      table: {
        category: 'Modifiers',
      },
    },
    kind: {
      table: {
        category: 'Modifiers',
      },
    },
    size: {
      table: {
        category: 'Modifiers',
      },
    },
    fullWidth: {
      table: {
        category: 'Modifiers',
      },
    },
    icon: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
    disabled: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
    loading: {
      type: '',
      table: {
        category: 'Modifiers',
      },
    },
  },
};

const Template: Story = (args) => (
  <div className="flex flex-col items-center justify-center space-y-8 h-full w-full">
    <Button {...args} />

    <Button {...args} color="success" />
  </div>
);

export const Default = Template.bind({});

export const WithIcon = Template.bind({});
WithIcon.args = {
  label: 'Add',
  icon: <FiPlus></FiPlus>,
};

export const IconOnly = Template.bind({});
IconOnly.args = {
  ariaLabel: 'Add',
  icon: <FiPlus></FiPlus>,
  variant: 'icon',
};
