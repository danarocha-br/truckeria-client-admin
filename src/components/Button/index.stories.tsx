import { Button, IconButton } from '@chakra-ui/react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { FiEdit2 } from 'react-icons/fi';

export default {
  title: 'Components/Button',
  component: Button,
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: ['sm', 'md'],
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['primary', 'transparent'],
      },
    },
  },
  args: {
    size: 'md',
    isActive: false,
    isDisabled: false,
    isFullWidth: false,
    isLoading: false,
    variant: 'primary',
  },
} as Meta;

const Template: Story = (args) => <Button {...args}>I am a Label</Button>;

export const Primary = Template.bind({});

export const IconButtonDefault: Story = (args) => (
  <IconButton
    {...args}
    aria-label="action-name"
    variant="transparent"
    icon={<FiEdit2 />}
  ></IconButton>
);
