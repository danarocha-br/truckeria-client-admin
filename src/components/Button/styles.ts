import styled, { css, DefaultTheme } from 'styled-components';
import tw from 'twin.macro';

import { ButtonProps } from './';

type ButtonStyledProps = {
  hasIcon: boolean;
} & Omit<ButtonProps, 'label'>;

const sizeModifier = {
  sm: (variant: ButtonProps) => css`
    ${tw`
        text-sm
        font-bold
        leading-4
        h-10
        rounded-xs
    `};

    ${variant === 'icon' ? tw`w-10` : tw`px-10`};

    svg {
      ${tw`
        h-6
        w-6
    `};
    }
  `,

  md: (variant: ButtonProps) => css`
    ${tw`
        text-base
        leading-5
        h-12
        rounded-xs
    `}
    ${variant === 'icon' ? tw`w-14` : tw`py-10 px-16`};

    svg {
      ${tw`
        h-7
        w-7
    `};
    }
  `,

  lg: (variant: ButtonProps) => css`
    ${tw`
        text-lg
        leading-6
        h-16
        rounded-sm
    `}
    ${variant === 'icon' ? tw`w-16` : tw`py-12 px-20`};

    svg {
      ${tw`
        h-8
        w-8
    `};
    }
  `,
};

const iconModifier = {
  hasIcon: () => css`
    svg {
      ${tw`
        mr-4
        rounded-md
        w-6
        h-6
        -ml-4
      `}
    }
  `,
};

const colorModifier = {
  primary: (theme: DefaultTheme) => css`
    background-color: ${theme.action.primary.default};
    ${tw`focus:ring-brand`}

    &:hover {
      background-color: ${theme.action.primary.hover};
    }
    &:focus {
      background-color: ${theme.action.primary.hover};
      box-shadow: 0 0 0 1px ${theme.background.default},
        0 0 0 calc(2px + 2px) ${theme.action.primary.disabled};
    }
    &:active {
      background-color: ${theme.action.primary.pressed};
    }

    &:disabled {
      background-color: ${theme.action.primary.disabled};
    }
  `,

  success: (theme: DefaultTheme) => css`
    background-color: ${theme.action.success.default};

    &:hover {
      background-color: ${theme.action.success.hover};
    }
    &:focus {
      background-color: ${theme.action.success.hover};
      box-shadow: 0 0 0 1px ${theme.background.default},
        0 0 0 calc(2px + 2px) ${theme.action.success.disabled};
    }
    &:active {
      background-color: ${theme.action.success.pressed};
    }

    &:disabled {
      background-color: ${theme.action.success.disabled};
    }
  `,
};

export const Button = styled.button<ButtonStyledProps>`
  ${({ theme, size, fullWidth, hasIcon, color, variant }) => css`
    ${tw`
        text-white
        font-semibold
        text-center
        flex
        items-center
        justify-center
        whitespace-nowrap
        w-auto
        relative
        transition-all
        duration-300

        focus:outline-none
        focus:ring-4
    `}
    ${fullWidth && tw`w-full`}

    ${hasIcon && iconModifier.hasIcon()}
    ${!!size && sizeModifier[size](variant)}
    ${!!color && colorModifier[color](theme)}

    svg {
      ${variant === 'icon' && tw`ml-0 mr-0`};
    }
  `}
`;
