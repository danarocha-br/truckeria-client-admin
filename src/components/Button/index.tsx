import React from 'react';
import { CSS } from '@stitches/react';

import { Spinner } from '../Spinner';
import { Icon, iconPath } from '../Icon';
import * as S from './styles';
import { config } from '../../../stitches.config';

export type ButtonProps = {
  label?: string;
  ariaLabel?: string;
  icon?: keyof typeof iconPath;
  variant?: 'primary' | 'flat' | 'success' | 'icon' | 'danger';
  size?: 'sm' | 'md' | 'lg';
  fullWidth?: boolean;
  loading?: boolean;
  disabled?: boolean;
  css?: CSS<typeof config>;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

export const Button = ({
  label,
  ariaLabel,
  variant = 'primary',
  size = 'md',
  fullWidth = false,
  loading = false,
  disabled = false,
  icon,
  onClick,
  ...props
}: ButtonProps) => (
  <S.Container
    aria-label={ariaLabel ? ariaLabel : label}
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon && variant !== 'icon'}
    variant={variant}
    loading={loading}
    disabled={disabled || loading}
    onClick={onClick}
    {...props}
  >
    {!!icon && <Icon name={icon} color="current" />}
    {variant !== 'icon' && label}
    {loading && variant !== 'icon' && (
      <S.LoadingWrapper>
        <Spinner size="xs" color="onInteractive" />
      </S.LoadingWrapper>
    )}
  </S.Container>
);
