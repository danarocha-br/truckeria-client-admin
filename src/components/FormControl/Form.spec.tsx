import { renderWithTheme } from 'utils/tests/helpers';

import Form from '.';
import Button from 'components/Button';

describe('<Form />', () => {
  it('should render the Form', () => {
    const { container } = renderWithTheme(
      <Form onSubmit={() => console.log('submit')} data-testid="form">
        <input type="text" />
        <Button label="Send" color="success" type="submit"></Button>
      </Form>
    );

    expect(container.querySelector('form')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
