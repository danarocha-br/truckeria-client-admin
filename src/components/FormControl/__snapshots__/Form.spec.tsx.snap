// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render the Form 1`] = `
.c0 {
  --tw-text-opacity: 1;
  color: rgba(255,255,255,var(--tw-text-opacity));
  font-weight: 600;
  text-align: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  width: auto;
  position: relative;
  -webkit-transition-property: all;
  transition-property: all;
  -webkit-transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  transition-timing-function: cubic-bezier(0.4,0,0.2,1);
  -webkit-transition-duration: 300ms;
  transition-duration: 300ms;
  font-size: 1.1rem;
  line-height: 1.25rem;
  height: 3rem;
  border-radius: 0.4rem;
  padding-top: 2.5rem;
  padding-bottom: 2.5rem;
  padding-left: 4rem;
  padding-right: 4rem;
  background-color: #76C248;
}

.c0:focus {
  outline: 2px solid transparent;
  outline-offset: 2px;
  --tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);
  --tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(4px + var(--tw-ring-offset-width)) var(--tw-ring-color);
  box-shadow: var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow,0 0 #0000);
}

.c0 svg {
  height: 1.75rem;
  width: 1.75rem;
}

.c0:hover {
  background-color: #5ea136;
}

.c0:focus {
  background-color: #5ea136;
  box-shadow: 0 0 0 1px #FFFFFF,0 0 0 calc(2px + 2px) #829179;
}

.c0:active {
  background-color: #487b29;
}

.c0:disabled {
  background-color: #829179;
}

<form
  class=" w-full space-y-5 space undefined"
>
  <input
    type="text"
  />
  <button
    aria-label="Send"
    class="c0"
    color="success"
    loading="0"
    type="submit"
  >
    Send
  </button>
</form>
`;
